core/src/engine/impl/mutex_impl.hpp

```diff
@@ -0,0 +1,167 @@
+ public:
+  MutexWaitStrategy(MutexImpl<WaitList>& mutex, TaskContext& current,
+                    Deadline deadline)
+      : WaitStrategy(deadline),
+        mutex_(mutex),
+        current_(current),
+        waiter_token_(mutex_.lock_waiters_),
+        lock_(mutex_.lock_waiters_) {}
+
+  void SetupWakeups() override {
+    mutex_.lock_waiters_.Append(lock_, &current_);
+    lock_.unlock();
+  }
+
+  void DisableWakeups() override {
+    lock_.lock();
+    mutex_.lock_waiters_.Remove(lock_, current_);
+  }
+
+ private:
+  MutexImpl<WaitList>& mutex_;
+  TaskContext& current_;
+  const WaitList::WaitersScopeCounter waiter_token_;
+  WaitList::Lock lock_;
+};
+
+template<>
+class MutexImpl<WaitListLight>::MutexWaitStrategy final : public WaitStrategy {
+ public:
+  MutexWaitStrategy(MutexImpl<WaitListLight>& mutex, TaskContext& current,
+                    Deadline deadline)
+      : WaitStrategy(deadline),
+        mutex_(mutex),
+        current_(current) {}
+
+  void SetupWakeups() override {
+    mutex_.lock_waiters_.Append(&current_);
+    if (!mutex_.owner_.load()) mutex_.lock_waiters_.WakeupOne();
+  }
+
+  void DisableWakeups() override {
+    mutex_.lock_waiters_.Remove(current_);
+  }
+
+ private:
+  MutexImpl<WaitListLight>& mutex_;
+  TaskContext& current_;
+};
+
+template <class Waiters>
+MutexImpl<Waiters>::MutexImpl() : owner_(nullptr) {}
+
+template <class Waiters>
+MutexImpl<Waiters>::~MutexImpl() { UASSERT(!owner_); }
+
+template <class Waiters>
+bool MutexImpl<Waiters>::LockFastPath(TaskContext& current) {
+  TaskContext* expected = nullptr;
+  return owner_.compare_exchange_strong(expected, &current,
+                                        std::memory_order_acquire);
+}
+
+template <class Waiters>
+bool MutexImpl<Waiters>::LockSlowPath(TaskContext& current, Deadline deadline) {
+  TaskContext* expected = nullptr;
+
+  engine::TaskCancellationBlocker block_cancels;
+  MutexWaitStrategy wait_manager(*this, current, deadline);
+  while (!owner_.compare_exchange_strong(expected, &current,
+                                         std::memory_order_relaxed)) {
+    UINVARIANT(expected != &current,
+               "MutexImpl is locked twice from the same task");
+
+    if (current.Sleep(wait_manager) ==
+        TaskContext::WakeupSource::kDeadlineTimer) {
+      return false;
+    }
+
+    expected = nullptr;
+  }
+  return true;
+}
+
+template <class Waiters>
+void MutexImpl<Waiters>::lock() { try_lock_until(Deadline{}); }
+
+template <class Waiters>
+void MutexImpl<Waiters>::unlock() {
+  auto* old_owner = owner_.exchange(nullptr, std::memory_order_acq_rel);
+  UASSERT(old_owner && old_owner->IsCurrent());
+
+  if constexpr (std::is_same<Waiters, WaitList>::value) {
+    if (lock_waiters_.GetCountOfSleepies()) {
+      WaitList::Lock lock(lock_waiters_);
+      lock_waiters_.WakeupOne(lock);
+    }
+  }
+
+  if constexpr (std::is_same<Waiters, WaitListLight>::value) {
+    lock_waiters_.WakeupOne();
```

Comment:
I'm not entirely sure we can get away with only acquire-release ordering here (`WaitList::Lock` provides sequential order).

Reply:
As I understand, this issue is mentioned here

```
To account for `WakeupOne()` calls between condition check and
  `Sleep` + `Append`, you have to recheck the condition after `Append`
   returns in `SetupWakeups`.
```
that's why

```
void SetupWakeups() override {
    mutex_.lock_waiters_.Append(&current_);
    if (!mutex_.owner_.load()) mutex_.lock_waiters_.WakeupOne();
  }
```

Reply:
Not really, I'm more afraid of data races in user code because of differences in behavior between our mutexes and standard ones. I have a sneaking suspicion that even something like AB-BA may become possible but no scenario to present as of yet.

I.e. [[thread.mutex.requirements.mutex.general]](http://eel.is/c++draft/thread.mutex.requirements.mutex.general) says 
> The lock and unlock operations on a single mutex appears to occur in a single total order.

And we're already not providing this guarantee in `lock()`.

Reply:
Frankly speaking, I do not clearly understand the matter of your suspicions. 

Lock, which is been using in Mutex implementation required because WaitList is not thread-safe. And as I understand, locks and unlock operations are made outside the object methods only for optimization purposes.
Let's see different scenarios.
1. One thread. Nothing can be wrong  In this case because WaitList will be empty all time.
2. Main case - 2 threads.
We have two points of synchronization: atomic MutexImpl.owner and atomic WaitListLight::Impl::waiter.

The first thread takes over the mutex. MutexImpl.owner equals first task context id.

If it released the lock, nothing would happen, because WaitListLight::Impl::waiter is empty.

The second thread tries to take over the mutex. It fails because MutexImpl::owner is not empty. 

If the first thread released the lock, nothing would happen, because WaitListLight::Impl::water is still empty.

The second thread adds itself to the WaitListLight::Impl::waiter. 

If the first thread released the lock, WakeupOne for the second thread would be called. Flag kWakeupByWaitList would be set for TaskContext::sleep_state_. But nothing would happen more, because the second thread steel is in a running state. 

The second thread checks that MutexImpl::owner is not empty. If it was empty WakeupOne for the second thread would be called. Flag kWakeupByWaitList would be set for TaskContext::sleep_state_. But nothing would happen more, because the second thread steel is in a running state. 

If the first thread released the lock, WakeupOne for the second thread would be called. Flag kWakeupByWaitList would be set for TaskContext::sleep_state_. But nothing would happen more, because the second thread steel is in a running state. 

The second thread has slept. If flag kWakeupByWaitList has been set, it would be scheduled.  

If the first thread released the lock, WakeupOne for the second thread would be called. Flag kWakeupByWaitList would be set for TaskContext::sleep_state_.  And second thread would be scheduled.

The second thread has slept. If flag kWakeupByWaitList has been set, it would be scheduled. 

If the first thread released the lock, WakeupOne for the second thread would be called. Flag kWakeupByWaitList would be set for TaskContext::sleep_state_. And second thread would be scheduled. 

The second thread wakeup. 

If the first thread released the lock, WakeupOne for the second thread would be called. Flag kWakeupByWaitList would be set for TaskContext::sleep_state_. But nothing would happen more, because the second thread steel is in a running state. 

The second thread removes itself to the WaitListLight::Impl::waiter.

If the first thread released the lock, nothing would happen, because WaitListLight::Impl::water is empty. 

The first thread can take over mutex again at this particular moment. This vulnerability exists in all mutex implementation in uServer. 

The second thread tries to take over the mutex. And we go on next iteration in case it fails.

Reply:
That test shows that we do not provide a guaranteed sequence lock and unlock for all mutex types.

 
```
TYPED_UTEST_P_MT(Mutex, LockUnlockOrder, 3) {
  static constexpr auto kTestDuration = std::chrono::milliseconds{500};

  const auto test_deadline = engine::Deadline::FromDuration(kTestDuration);
  const size_t worker_count = 2;
  const size_t iteration_count = 1000;

  TypeParam mutex;
  size_t error_count = 0;
  while (!test_deadline.IsReached()) {
    std::vector<engine::TaskWithResult<void>> tasks;
    size_t run_worker_count = 0;
    auto prev_begin_lock_time = std::chrono::steady_clock::now();
    for (size_t i = 0; i < worker_count; ++i) {
      const auto work = [&] {
        {
          std::unique_lock lock(mutex);
          ++run_worker_count;
        }
        while (true) {
          std::unique_lock lock(mutex);
          if (run_worker_count == worker_count) {
            break;
          }
        }
        for (size_t j = 0; j < iteration_count; ++j) {
          auto begin_lock = std::chrono::steady_clock::now();
          mutex.lock();
          auto diff = begin_lock - prev_begin_lock_time;
          if (std::chrono::duration_cast<std::chrono::nanoseconds>(diff).count() < -40000) {
            ++error_count;
          }
          prev_begin_lock_time = begin_lock;
          mutex.unlock();
        }
      };
      tasks.push_back(engine::AsyncNoSpan(work));
    }
    for (auto& task : tasks) task.Get();
  }
  EXPECT_EQ(error_count, 0u);
}
```

```
./userver/core/userver-core_unittest --gtest_filter=*LockUnlockOrder
Note: Google Test filter = *LockUnlockOrder
[==========] Running 3 tests from 3 test suites.
[----------] Global test environment set-up.
[----------] 1 test from EngineMutex/Mutex/0, where TypeParam = userver::engine::Mutex
[ RUN      ] EngineMutex/Mutex/0.LockUnlockOrder
/Users/kor-den/github/userver/core/src/engine/mutex_test.cpp:194: Failure
Expected equality of these values:
  error_count
    Which is: 46
  0u
    Which is: 0
[  FAILED  ] EngineMutex/Mutex/0.LockUnlockOrder, where TypeParam = userver::engine::Mutex (501 ms)
[----------] 1 test from EngineMutex/Mutex/0 (501 ms total)

[----------] 1 test from EngineSharedMutex/Mutex/0, where TypeParam = userver::engine::SharedMutex
[ RUN      ] EngineSharedMutex/Mutex/0.LockUnlockOrder
/Users/kor-den/github/userver/core/src/engine/mutex_test.cpp:194: Failure
Expected equality of these values:
  error_count
    Which is: 33
  0u
    Which is: 0
[  FAILED  ] EngineSharedMutex/Mutex/0.LockUnlockOrder, where TypeParam = userver::engine::SharedMutex (501 ms)
[----------] 1 test from EngineSharedMutex/Mutex/0 (501 ms total)

[----------] 1 test from EngineSingleWaitingTaskMutex/Mutex/0, where TypeParam = userver::engine::SingleWaitingTaskMutex
[ RUN      ] EngineSingleWaitingTaskMutex/Mutex/0.LockUnlockOrder
/Users/kor-den/github/userver/core/src/engine/mutex_test.cpp:194: Failure
Expected equality of these values:
  error_count
    Which is: 14
  0u
    Which is: 0
[  FAILED  ] EngineSingleWaitingTaskMutex/Mutex/0.LockUnlockOrder, where TypeParam = userver::engine::SingleWaitingTaskMutex (500 ms)
[----------] 1 test from EngineSingleWaitingTaskMutex/Mutex/0 (500 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 3 test suites ran. (1503 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] EngineMutex/Mutex/0.LockUnlockOrder, where TypeParam = userver::engine::Mutex
[  FAILED  ] EngineSharedMutex/Mutex/0.LockUnlockOrder, where TypeParam = userver::engine::SharedMutex
[  FAILED  ] EngineSingleWaitingTaskMutex/Mutex/0.LockUnlockOrder, where TypeParam = userver::engine::SingleWaitingTaskMutex

```
 3 FAILED TESTS